cleverhans/attacks.py

```diff
@@ -1,4 +1,342 @@
+from abc import ABCMeta
 import numpy as np
+import warnings
+
+
+class Attack:
+    """
+    Abstract base class for all attack classes.
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, model, back='tf', sess=None, params={}):
+        """
+        :param pred: A function that takes a symbolic input and returns the
+                     symbolic output for the model's predictions.
+        :param back: The backend to use. Either 'tf' (default) or 'th'.
+        :param params: Parameter dictionary used by child classes.
+        """
+        if not(back == 'tf' or back == 'th'):
+            raise Exception("Backend argument must either be 'tf' or 'th'.")
+        if back == 'tf' and sess is None:
+            raise Exception("A tf session was not provided in sess argument.")
+        if back == 'th' and sess is not None:
+            raise Exception("A session should not be provided when using th.")
+
+        # Prepare attributes
+        self.model = model
+        self.back = back
+        self.sess = sess
+
+    def generate(self, x, params={}):
+        """
+        Generate the attack's symbolic graph for adversarial examples. This
+        method should be overriden in any child class that implements an
+        attack that is expressable symbolically. Otherwise, it will wrap the
+        numerical implementation as a symbolic operator.
+        :param x: The model's symbolic inputs.
+        :param params: Parameter dictionary used by child classes.
+        :return: A symbolic representation of the adversarial examples.
+        """
+        if self.back == 'th':
+            raise NotImplementedError('Theano version not implemented.')
+
+        import tensorflow as tf
+        return tf.py_func(self.generate_np, [x], tf.float32)
```

Comment:
This needs to check to make sure it's not caught in an infinite loop with `generate` calling `generate_np` and vice versa. I recommend adding a unit test that checks that a subclass that implements neither method causes a NotImplementedError rather than an infinite loop.